Adds a mutating admission controller to inject GPU functionality

ellis-bigelow opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #17, #142

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Enables GPU Selection on GKE
Enables GPUs with Tensorflow
Enables GPUs with XGBoost

```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kfserving/143)
<!-- Reviewable:end -->

yuzisun commented on pkg/webhook/admission/deployment/accelerator_injector.go:
```diff
@@ -0,0 +1,91 @@
+	return nil
+}
+
+func injectGPUToleration(deployment *appsv1.Deployment) error {
```
Is this needed ? Not sure if this is standard way but we are using https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#extendedresourcetoleration to automatically add the toleration.

ellis-bigelow replied:
We're using the same. I wanted a catchall for people not using it. Happy to remove if this is standard. What do you think?

ellis-bigelow commented:
@cliveseldon 
@yuzisun 
This change is ready!

animeshsingh commented on docs/samples/accelerators/tensorflow-gpu.yaml:
```diff
@@ -0,0 +1,14 @@
+      runtimeVersion: "1.11.0-gpu"
+      resources:
+        limits:
+          nvidia.com/gpu: 1
```
what`s the rational for using nvidia.com here?

ellis-bigelow replied:
s/rational/rationale :)

This is the resource key name standardized by nvidia * k8s:
https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/#v1-8-onwards

animeshsingh commented on pkg/reconciler/knative/resources/knative_configuration_test.go:
```diff
@@ -33,14 +33,17 @@ var kfsvc = &v1alpha1.KFService{
 		Name:      "mnist",
 		Namespace: "default",
+		Annotations: map[string]string{
+			constants.KFServiceGKEAcceleratorAnnotationKey: "nvidia-tesla-t4",
```
Idea here i believe is to extend with annotations from other cloud providers if needed?

ellis-bigelow replied:
Absolutely. It's not clear to me if there are additional bits though

yuzisun commented on pkg/apis/serving/v1alpha1/kfservice_validation.go:
```diff
@@ -82,10 +80,6 @@ func validateModelSpec(spec *ModelSpec) error {
-	container := spec.CreateModelServingContainer("any", &FrameworksConfig{})
-	if err := knserving.ValidateContainer(*container, sets.String{}); err != nil {
-		return fmt.Errorf("container validation error: %s", err.Error())
-	}
```
+1 to delete this awkward code :),  I guess this is not needed since controller creates the container except custom model spec.

yuzisun commented on pkg/apis/serving/v1alpha1/framework_tensorflow_test.go:
```diff
@@ -0,0 +1,61 @@
+	resource "k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestAcceptGoodRuntimeVersion(t *testing.T) {
```
Should we follow the same testing style across the project? I think the best practice in golang for unit test is to use table testing (https://github.com/golang/go/wiki/TableDrivenTests).

ellis-bigelow replied:
Yeah I'll update these.

yuzisun commented on pkg/webhook/webhook.go:
```diff
@@ -111,5 +113,26 @@ func register(manager manager.Manager, server *webhook.Server) error {
 				Decoder: manager.GetAdmissionDecoder(),
 			},
 		},
+	}, &admission.Webhook{
```
Can we make this mutation webhook optional as I do not see it is needed for on prem use case ?

ellis-bigelow replied:
It shouldn't hurt, right?

yuzisun replied:
I am a bit worried about the number of admission webhooks now which adds the overhead of launching deployments on cluster.

hougangliu commented on pkg/webhook/webhook.go:
```diff
@@ -111,5 +113,26 @@ func register(manager manager.Manager, server *webhook.Server) error {
+			},
+		}},
+		Handlers: []admission.Handler{
+			&deployment.Mutator{
```
I don't think we should mutate all deployment is cluster. At least, we should not mutate deployments which is not related with kfserving (maybe we can filter by labels)

ellis-bigelow replied:
Agreed. I'll update this.

hougangliu replied:
I am afraid we have to do it in the handler so far.
But in long term, we expect that knative configuration can add new 'field' to support customize deployment fileds such as .Spec.NodeSelector and .Spec.Tolerations in this case. Have no idea if knative serving has such plan

hougangliu commented on pkg/webhook/admission/deployment/accelerator_injector.go:
```diff
@@ -0,0 +1,102 @@
+					Value:    NvidiaGPUTaintValue,
+					Operator: v1.TolerationOpEqual,
+					Effect:   v1.TaintEffectPreferNoSchedule,
+				})
```
the gpu taint is starndard? or just works for gcp platform?
BTW, should we append it only when it is not included in .Spec.Tolerations

ellis-bigelow replied:
I've dropped this in favor of the standard one described in the readme.

ellis-bigelow commented:
Hey All.
I've addressed comments with one outstanding: Disable Webhook option.

I really want to avoid forcing users to explicitly set this in a configmap. 

I believe that the controllerruntime SIG is working on a longer term solution to this to allow for filtering on Resource Labels/Annotations rather than just Namespace labels.

Is it alright to leave this as is? I think the overhead is pretty low.

yuzisun commented:
> Hey All.
> I've addressed comments with one outstanding: Disable Webhook option.
> 
> I really want to avoid forcing users to explicitly set this in a configmap.
> 
> I believe that the controllerruntime SIG is working on a longer term solution to this to allow for filtering on Resource Labels/Annotations rather than just Namespace labels.
> 
> Is it alright to leave this as is? I think the overhead is pretty low.

Filtering by Resource Labels/Annotations would be a great feature! I am fine for now but if we are adding more and more webhooks I'd hope we can deploy webhook separately instead of embedding in our controller and people can pick the ones they need.

yuzisun commented on pkg/webhook/admission/deployment/accelerator_injector.go:
```diff
@@ -0,0 +1,81 @@
+	return third_party.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, patch)
+}
+
+func Mutate(deployment *appsv1.Deployment) error {
```
nit: should this be private ?

yuzisun commented:
/lgtm

yuzisun commented:
Looks all good !
/lgtm
/approve
